lib/sycamore/sycamore/transforms/table_structure/extract.py

```diff
@@ -0,0 +1,143 @@
+from torchvision import transforms
+import torch
+
+from sycamore.data import BoundingBox, Element, Document, TableElement
+from sycamore.plan_nodes import Node, Transform
+from sycamore.transforms.table_structure import table_transformers
+from sycamore.transforms.table_structure.table_transformers import MaxResize
+from sycamore.utils.generate_ray_func import generate_map_function
+
+
+class TableStructureExtractor:
+    @abstractmethod
+    def extract(self, element: TableElement, doc_image: Image) -> TableElement:
+        pass
+
+    def extract_from_doc(self, doc: Document) -> Document:
+        """Method that extracts the table structure for each table in the Document.
+
+        Suitable for use in a map function. The binary_representation of the doc
+        should be the pdf bytes.
+
+        This method is best effort. If a table element is missing required metadata
+        it will be skipped and no error will be thrown.
+        """
+
+        # TODO: Perhaps we should support image formats in addition to PDFs.
+        if doc.binary_representation is None:
+            return doc
+
+        images = pdf2image.convert_from_bytes(doc.binary_representation)
+        new_elements: list[Element] = []
+
+        for elem in doc.elements:
+            if isinstance(elem, TableElement):
+                if "page_number" in elem.properties:
+                    page_num = elem.properties["page_number"] - 1
+                elif len(images) == 1:
+                    page_num = 0
+                else:
+                    new_elements.append(elem)
+                    continue
+                new_elements.append(self.extract(elem, images[page_num]))
+            else:
+                new_elements.append(elem)
+        doc.elements = new_elements
+        return doc
+
+
+class TableTransformerStructureExtractor(TableStructureExtractor):
+    DEFAULT_TTAR_MODEL = "microsoft/table-structure-recognition-v1.1-all"
+
+    def __init__(self, model: str = DEFAULT_TTAR_MODEL):
+        self.model = model
+
+    # Convert tokens (text) into the format expected by the TableTransformer
+    # postprocessing code.
+    def _prepare_tokens(self, tokens: list[dict[str, Any]], crop_box, width, height) -> list[dict[str, Any]]:
+        for i, t in enumerate(tokens):
+            assert isinstance(t["bbox"], BoundingBox)
+            t["bbox"].to_absolute_self(width, height).translate_self(-1 * crop_box[0], -1 * crop_box[1])
+            t["bbox"] = t["bbox"].to_list()
+            t["span_num"] = i
+            t["line_num"] = 0
+            t["block_num"] = 0
+        return tokens
+
+    def extract(self, element: TableElement, doc_image: Image.Image) -> TableElement:
+        # We need a bounding box to be able to do anything.
+        if element.bbox is None:
+            return element
+
+        width, height = doc_image.size
+
+        structure_model = TableTransformerForObjectDetection.from_pretrained(self.model)
```

Comment:
In this case, we are reloading the model each time this called for a table?

Reply:
Ah yes, you are right. I see how we could change it to load once per document, which I will do. Do you think there's a better way to do it?

Reply:
We discuss a bit before on partitioner, I feel the best way would be put in actor class and only call once per worker.

Reply:
I want to make sure I understand the proposal here. Are you suggesting we convert the partition transform to use a map class instead of a function for the callable? And then I guess the init for that would initialize the partitioners and any models. Is that right?

Reply:
yes, that's what I mean. This would change the docset partition API, we pass extractor parameters instead of extractor instance. If you have concern on this, I'm good with document level model initialization.

Reply:
I don't have any major objections to this. I am wondering if we want to do it in this PR or do a followup. There is some urgency for this change b/c we want to get a blog post out, but I don't want to steamroll over your comments. I can sync with you today.